### PR TITLE
Connect remove account rpc request

### LINKED ISSUE #18422: Wallet - Connect RPC for removing account
# Feature Issue

Connect `status-go` RPC for removing wallet accounts.

### Description

- On confirmation, the RPC should be called with the address of the account to be removed
- On successful removal, the toast with the message `Account has been removed` (`Watched address has been removed` for watched address) is shown
- Call the `:wallet/get-accounts` event upon successful removal to refresh the accounts list

### Acceptance Criteria

- Toast message is shown upon successful removal
- The account is removed successfully with no listings in Wallet home, Account Switcher,...etc.
- The total accounts balance should reflect to the new list of accounts

### Notes

The default account can not be removed.

### PR DESCRIPTION
**Pull request by Rende11**
Fixes #18422

### Summary
This PR connects rpc request to delete account. 
After successful deletion, user redirects to accounts list, a toast is shown, and the list of accounts is updated.

Also remove options are hidden for default-account

### Testing notes
Now it's hidden under feature-flag
Removing feature is unavailable for already created generated accounts!

https://github.com/status-im/status-mobile/assets/10757633/b81c099e-c354-4d16-b3a3-4bcc6c57e7b8

status: ready

### CODE REVIEW

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs [hunk 1]
```diff
@@ -32,12 +32,13 @@
...
-                                                        network-preference-details
-                                                        address)
-        share-title                                    (str name " " (i18n/label :t/address))]
+  (let [{:keys            [name color emoji address watch-only?]
+         default-account? :wallet} (rf/sub [:wallet/current-viewing-account])
+        network-preference-details (rf/sub [:wallet/network-preference-details])
+        multichain-address         (utils/get-multichain-address
+                                    network-preference-details
+                                    address)
+        share-title                (str name " " (i18n/label :t/address))]
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: this is better as an i18n label with a param.
e.g  `(i18n/label :t/address {:name name})`

File: src/status_im/contexts/wallet/events.cljs [hunk 2]
```diff
@@ -84,6 +84,30 @@
                                    {:error %
                                     :event :wallet/save-account})}]]]}))
 
+(rf/reg-event-fx
+ :wallet/show-account-deleted-toast
+ (fn [_ [message]]
+   {:dispatch [:toasts/upsert {:type :positive :text message}]}))
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/events.cljs]: nest to align maps keys 👍

File: src/status_im/contexts/wallet/events.cljs [hunk 3]
```diff
@@ -133,10 +157,10 @@
 
 (rf/reg-event-fx :wallet/create-derived-addresses
  (fn [{:keys [db]} [{:keys [sha3-pwd path]} on-success]]
-   (let [{:keys [wallet-root-address]} (:profile/profile db)]
+   (let [{:keys [address]} (:profile/profile db)]
      {:fx [[:json-rpc/call
             [{:method     "wallet_getDerivedAddresses"
-              :params     [sha3-pwd wallet-root-address [path]]
+              :params     [sha3-pwd address [path]]
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/events.cljs:163]: 👍

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs [hunk 4]
```diff
@@ -79,13 +86,17 @@
    [rn/view {:style style/desc-container}
     [quo/text {:weight :medium}
      (i18n/label :t/remove-watched-address-desc)]]
-   [footer {:submit-disabled? false}]])
+   [footer
+    {:submit-disabled? false
+     :address          address
+     :toast            (i18n/label :t/watched-account-removed)}]])
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: consider :toast-label/ :toast message instead. 
Toasts are generally an object of data and I was more expecting to see that here

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs [hunk 5]
```diff
@@ -92,15 +93,16 @@
...
-        {:add-divider?        (not show-account-selector?)
-         :icon                :i/delete
-         :accessibility-label :remove-account
-         :label               (i18n/label :t/remove-account)
-         :danger?             true
-         :on-press            #(rf/dispatch [:show-bottom-sheet
-                                             {:content
-                                              (fn []
-                                                [remove-account/view])}])}]]]
+        (when (not default-account?)
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: Very small: we can use `when-not` :)

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs [hunk 6]
```diff
@@ -32,12 +32,13 @@
...
 (defn- options
   [{:keys [theme show-account-selector? options-height]}]
-  (let [{:keys [name color emoji address watch-only?]} (rf/sub [:wallet/current-viewing-account])
-        network-preference-details                     (rf/sub [:wallet/network-preference-details])
-        multichain-address                             (utils/get-multichain-address
-                                                        network-preference-details
-                                                        address)
-        share-title                                    (str name " " (i18n/label :t/address))]
+  (let [{:keys            [name color emoji address watch-only?]
+         default-account? :wallet} (rf/sub [:wallet/current-viewing-account])
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: Extracting the `:wallet` key to check if it's the default account or not seems error-prone. We might change the structure of the db and we won't notice this is breaking.

cc: @J-Son89 is exracting `:wallet` the proper way to check if it's the default account? is the`:created-at` key reliable to check this?

@Rende11 
I think the proper way to achieve it is adding a `:default-account?` key when we receive the accounts to store.

The event is `:wallet/get-accounts-success` in the ns `status-im.contexts.wallet.events`, it'd be really good if you check which is the default account and add the previously mentioned keyword, it'll make the info clearer and also availabe to all the subs

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs [hunk 7]
```diff
@@ -11,13 +11,17 @@
...
    {:actions             :2-actions
     :customization-color :danger
     :button-one-label    (i18n/label :t/remove)
-    :button-one-props    {:on-press  #(js/alert "Will be implemented")
-                          :type      :danger
+    :button-one-props    {:on-press
+                          (fn []
+                            (rf/dispatch [:wallet/remove-account
+                                          {:address address
+                                           :toast   toast}]))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: :+1:

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs [hunk 8]
```diff
@@ -79,13 +86,17 @@
...
+     :toast            (i18n/label :t/watched-account-removed)}]])
 
 (defn- view-internal
   []
   (let [{:keys [type] :as account} (rf/sub [:wallet/current-viewing-account])]
     (case type
       :generated [recovery-phase-flow account]
-      :watch     [watched-address-flow account])))
+      :watch     [watched-address-flow account]
+      nil)))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs:100]: :100:

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs [hunk 9]
```diff
@@ -64,10 +68,13 @@
             :checked?            @confirmed?
             :on-change           #(swap! confirmed? not)}]
           [quo/text (i18n/label :t/remove-account-confirmation)]]
-         [footer {:submit-disabled? (not @confirmed?)}]]))))
+         [footer
+          {:submit-disabled? (not @confirmed?)
+           :address          address
+           :toast            (i18n/label :t/account-removed)}]]))))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: Will the `footer` dispatch a toast? :thinking:

File: src/status_im/contexts/wallet/edit_account/view.cljs [hunk 10]
```diff
@@ -51,24 +51,26 @@
...
-            test-networks-enabled?  (rf/sub [:profile/test-networks-enabled?])
-            network-preferences-key (if test-networks-enabled?
-                                      :test-preferred-chain-ids
-                                      :prod-preferred-chain-ids)
-            account-name            (or @edited-account-name name)
-            button-disabled?        (or (nil? @edited-account-name)
-                                        (= name @edited-account-name))]
+      (let [{:keys            [name emoji address color watch-only?]
+             default-account? :wallet
+             :as              account} (rf/sub [:wallet/current-viewing-account])
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/edit_account/view.cljs]: this destructuring will need to be updated if we add the key previously mentioned. Just pointing it as a reminder

File: src/status_im/contexts/wallet/events.cljs [hunk 11]
```diff
@@ -51,10 +52,9 @@
          wallet-db           (get db :wallet)
          new-account?        (:new-account? wallet-db)
          navigate-to-account (:navigate-to-account wallet-db)]
-     {:db (reduce (fn [db {:keys [address] :as account}]
-                    (assoc-in db [:wallet :accounts address] account))
-                  db
-                  (data-store/rpc->accounts wallet-accounts))
+     {:db (assoc-in db
+           [:wallet :accounts]
+           (utils.collection/index-by :address (data-store/rpc->accounts wallet-accounts)))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs:57]: :+1: cleaner!

File: src/status_im/contexts/wallet/events.cljs (hunk 2 above)
Comment (ulisesmac) [reply src/status_im/contexts/wallet/events.cljs]: also the suggested approach is using `:fx` with :dispatch inside

**Review by ulisesmac — COMMENTED** [review]
Hey @Rende11 
It looks good! :+1: just left some comments

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs (hunk 1 above)
Comment (Rende11) [reply src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: It's not related to my issue, but I will fix it)))

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs (hunk 9 above)
Comment (Rende11) [reply src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: It's a toast message after successful deletion

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs (hunk 6 above)
Comment (J-Son89) [reply src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: @ulisesmac It is correct to check :wallet key. It is an unfortunately named key from status go 🥲
We probably should rename it when we store it in the db to :default-account?

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs (hunk 4 above)
Comment (Rende11) [reply src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: Replaced with `:toast-message`

File: src/status_im/contexts/wallet/common/sheets/account_options/view.cljs (hunk 6 above)
Comment (Rende11) [reply src/status_im/contexts/wallet/common/sheets/account_options/view.cljs]: I added `:default-account?` key - see src/status_im/contexts/wallet/data_store.cljs

File: src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs (hunk 9 above)
Comment (Rende11) [reply src/status_im/contexts/wallet/common/sheets/remove_account/view.cljs]: Replaced with `:toast-message`

File: src/quo/components/gradient/gradient_cover/view.cljs [hunk 12]
```diff
@@ -6,14 +6,21 @@
...
-        color-bottom (colors/custom-color customization-color 50 0)]
-    [linear-gradient/linear-gradient
-     {:accessibility-label :gradient-cover
-      :colors              [color-top color-bottom]
-      :start               {:x 0 :y 0}
-      :end                 {:x 0 :y 1}
-      :style               (merge (style/root-container opacity height) container-style)}]))
+  [{:keys [customization-color opacity container-style height]
+    :or   {customization-color :blue}}]
+  ;; `when` added for safety, `linear-gradient` will break if `nil` is passed,
```
Comment (J-Son89) [inline src/quo/components/gradient/gradient_cover/view.cljs:11]: why don't we just wrap the linear gradient in a `when` to do a nil check. Or better again do this directly in `react-native.linear-gradient`. We can then remove this comment as it is a generic solution 👍

File: src/status_im/contexts/wallet/edit_account/view.cljs [hunk 13]
```diff
@@ -62,13 +62,14 @@
...
                                         (= name @edited-account-name))]
         [create-or-edit-account/view
-         {:page-nav-right-side [{:icon-name :i/delete
-                                 :on-press
-                                 (fn []
-                                   (rf/dispatch [:show-bottom-sheet
-                                                 {:content
-                                                  (fn []
-                                                    [remove-account/view])}]))}]
+         {:page-nav-right-side [(when-not default-account?
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/edit_account/view.cljs:66]: there is one more issue in this feature epic - https://github.com/status-im/status-mobile/issues/18575

can we hide these on-presses behind a feature-flag for now? and then we will review all of this in go?

Comment (J-Son89) [reply src/status_im/contexts/wallet/edit_account/view.cljs:66]: see https://github.com/status-im/status-mobile/pull/18569/files#diff-f121c6f060282c0f68513d7962d77967ef0cd9b5312c19f6b59304fed19c0108 for an example of what I mean

File: src/quo/components/gradient/gradient_cover/view.cljs (hunk 12 above)
Comment (Rende11) [reply src/quo/components/gradient/gradient_cover/view.cljs:11]: I think it's better solve in separate issue

**Review by ulisesmac — APPROVED** [review]

**Review by J-Son89 — APPROVED** [review]

**Comment by J-Son89** [comment]
@Rende11 - can you check component spec tests??

**Comment by J-Son89** [comment]
@pavloburykh - this feature is behind a feature flag and should be safe with just e2e / a quick smoke test. 
What do you think?

**Comment by pavloburykh** [comment]
Hey @Rende11 @J-Son89 In case PR does not require manual qa - then it is ready for merge. Failed e2e are not PR related.

File: src/status_im/contexts/wallet/data_store.cljs [hunk 14]
```diff
@@ -49,7 +50,8 @@
                         :color                    :colorId})
       (update :prodPreferredChainIds chain-ids-set->string)
       (update :testPreferredChainIds chain-ids-set->string)
-      (dissoc :watch-only?)))
+      (dissoc :watch-only?)
+      (dissoc :default-account?)))
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/data_store.cljs:54]: 👍 
cc @smohamedjavid for reference we add/remove default-account value now

**Pull request merged**